spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+/**
+ * Table scan type. It allows to create a Relation for given table, keyspace, cluster and user
+ * defined schema. It also allow to pass [[CassandraConnector]], [[ReadConf]] and [[WriteConf]]
+ * explicitly
+ */
+sealed trait ScanType {
```

Comment:
Looking at how Spark uses those scans, there is no need to implement all of them. Is there anything that our Catalyst scan can do better than our PrunedFilteredScan? I guess no, so CatalystScan is not needed at all. Particularly I'm against using CatalystScan, as the docs says explicitly it is going to break binary compatibility.

Do we ever want to disable column pruning? No. The only thing that might be worth considering is disabling predicate pushdown. So, we only need to implement PrunedFilteredScan and add an option to disable predicate pushdown.